File: backend/app/constant/pumpfun.py
from app.constant.config import BITQUERY_TOKEN


BITQUERY_URL = "https://streaming.bitquery.io/eap"  
BITQUERY_HEADERS = {
    "Content-Type": "application/json",
    "Authorization": f"Bearer {BITQUERY_TOKEN}" 
}

TOP_HOLDERS_QUERY = """
query MyQuery($token: String!) {
  Solana(dataset: realtime) {
    BalanceUpdates(
      limit: { count: 10 }
      orderBy: { descendingByField: "BalanceUpdate_Holding_maximum" }
      where: {
        BalanceUpdate: {
          Currency: {
            MintAddress: { is: $token }
          }
        }
        Transaction: { Result: { Success: true } }
      }
    ) {
      BalanceUpdate {
        Currency {
          Name
          MintAddress
          Symbol
        }
        Account {
          Address
        }
        Holding: PostBalance(maximum: Block_Slot)
      }
    }
  }
}
"""

DEV_HOLDINGS_QUERY = """
query MyQuery($dev: String!, $token: String!) {
  Solana {
    BalanceUpdates(
      where: {
        BalanceUpdate: {
          Account: { Address: { is: $dev } }
          Currency: { MintAddress: { is: $token } }
        }
      }
    ) {
      BalanceUpdate {
        balance: PostBalance(maximum: Block_Slot)
      }
    }
  }
}
"""

HOLDER_DISTRIBUTION_QUERY="""
query HolderDistribution($token: String!) {
  Solana {
    BalanceUpdates(
      orderBy: { descendingByField: "BalanceUpdate_Holding_maximum" }
      where: {
        BalanceUpdate: {
          Currency: {
            MintAddress: { is: $token }
          }
        }
        Transaction: { Result: { Success: true } }
      }
    ) {
      BalanceUpdate {
        Account {
          Address
        }
        Holding: PostBalance(maximum: Block_Slot)
      }
    }
  }
}
"""

VOLUME_AND_MARKETCAP_QUERY="""
query MyQuery($time_1h_ago: DateTime, $token: String, $side: String) {
  Solana {
    liquidity: DEXPools(
      where: {Pool: {Market: {BaseCurrency: {MintAddress: {is: $token}}, QuoteCurrency: {MintAddress: {is: $side}}}, Dex: {ProtocolName: {is: "pump"}}}}
      limit: {count: 1}
      orderBy: {descending: Block_Time}
    ) {
      Pool {
        Base {
          PostAmountInUSD
        }
      }
    }
    marketcap: TokenSupplyUpdates(
      where: {TokenSupplyUpdate: {Currency: {MintAddress: {is: $token}}}, Block: {Time: {till: $time_1h_ago}}}
      limitBy: {by: TokenSupplyUpdate_Currency_MintAddress, count: 1}
      orderBy: {descending: Block_Time}
    ) {
      TokenSupplyUpdate {
        PostBalanceInUSD
        Currency {
          Name
          MintAddress
          Symbol
        }
      }
    }
  }
}
"""

TOP_MARKET_CAP_PUMPFUN_COIN = """
query TopMarketCapCoin {
  Solana {
    TokenSupplyUpdates(
      limit: {count: 10}
      orderBy: {descendingByField: "TokenSupplyUpdate_Marketcap"}
      limitBy: {by: TokenSupplyUpdate_Currency_MintAddress, count: 1}
      where: {TokenSupplyUpdate: {Currency: {MintAddress: {includes: "pump"}}}}
    ) {
      TokenSupplyUpdate {
        Marketcap: PostBalanceInUSD
        Currency {
          Name
          Symbol
          MintAddress
          Fungible
          Decimals
        }
      }
    }
  }
}
"""

GET_TOKEN_INFORMATION = """
query TokenTradesInfo($token: String!, $before_timestamp: DateTime!) {
  Solana(dataset: realtime) {
    DEXTradeByTokens(
      where: {
        Transaction: {Result: {Success: true}}, 
        Trade: {
          Currency: {MintAddress: {is: $token}}
        }, 
        Block: {Time: {since: $before_timestamp}}}
    ) {
      Trade {
        Currency {
          Name
          MintAddress
          Symbol
        }
        start: PriceInUSD(minimum: Block_Time)
        end: PriceInUSD(maximum: Block_Time)
        Dex {
          ProtocolName
          ProtocolFamily
          ProgramAddress
        }
        Market {
          MarketAddress
        }
        Side {
          Currency {
            Symbol
            Name
            MintAddress
          }
        }
      }
      num_makers: count(distinct: Transaction_Signer)
      num_buyers: count(
        distinct: Transaction_Signer
        if: {Trade: {Side: {Type: {is: buy}}}}
      )
      num_sellers: count(
        distinct: Transaction_Signer
        if: {Trade: {Side: {Type: {is: sell}}}}
      )
      num_trades: count
      traded_volume: sum(of: Trade_Side_AmountInUSD)
      buy_volume: sum(
        of: Trade_Side_AmountInUSD
        if: {Trade: {Side: {Type: {is: buy}}}}
      )
      sell_volume: sum(
        of: Trade_Side_AmountInUSD
        if: {Trade: {Side: {Type: {is: sell}}}}
      )
      num_buys: count(if: {Trade: {Side: {Type: {is: buy}}}})
      num_sells: count(if: {Trade: {Side: {Type: {is: sell}}}})
    }
  }
}
"""

# Top Token Creators on Pump Fun
TOP_TOKEN_CREATORS_PUMPFUN_QUERY= """
query MyQuery {
  Solana(network: solana) {
    Instructions(
      where: {Instruction: {Program: {Name: {is: "pump"}, Method: {is: "create"}}}}
      orderBy: {descendingByField: "tokens_count"}
    ) {
      tokens_count: count
      Transaction {
        Signer
      }
    }
  }
}
"""

# on pumpfun dex
GET_TOP_TRADER_TOKEN_PUMPFUN_DEX_QUERY = """
query TopTraders($token: String, $limit: Int) {
  Solana {
    DEXTradeByTokens(
      orderBy: {descendingByField: "volumeUsd"}
      limit: {count: $limit}
      where: {Trade: {Currency: {MintAddress: {is: $token}}}, Transaction: {Result: {Success: true}}}
    ) {
      Trade {
        Account {
          Owner
        }
        Side {
          Account {
            Address
          }
          Type
        }
      }
      bought: sum(of: Trade_Amount, if: {Trade: {Side: {Type: {is: buy}}}})
      sold: sum(of: Trade_Amount, if: {Trade: {Side: {Type: {is: sell}}}})
      volume: sum(of: Trade_Amount)
      volumeUsd: sum(of: Trade_Side_AmountInUSD)
    }
  }
}
"""

# on multiple dex, return a list of volume of a token on each dex
GET_TRADING_VOLUME_TOKEN_QUERY = """
query MyQuery($token: String!, $since_time: DateTime!) {
  Solana {
    DEXTradeByTokens(
      where: {
        Trade: {
          Currency: {
            MintAddress: { is: $token }
          }
        }
        Block: { Time: { since: $since_time } }
      }
    ) {
      Trade {
        Currency {
          Name
          Symbol
          MintAddress
        }
        Dex {
          ProtocolName
          ProtocolFamily
        }
      }
      TradeVolume: sum(of: Trade_Amount)
    }
  }
}
"""


GET_FIST_BUYERS_PUMPFUN_TOKEN_QUERY = """
query MyQuery($token: String!, $limit: Int) {
  Solana {
    DEXTrades(
      where: {
        Trade: {
          Buy: {
            Currency: {
              MintAddress: {
                is: $token
              }
            }
          }
        }
      }
      limit: { count: $limit }
      orderBy: { ascending: Block_Time }
    ) {
      Trade {
        Buy {
          Amount
          Account {
            Token {
              Owner
            }
          }
        }
      }
    }
  }
}
"""

PUMPFUN_TOKEN_LATEST_TRADES_QUERY ="""
query pumpfunTokenLatestTrades($token: String, $limit: Int) {
  Solana {
    DEXTradeByTokens(
      orderBy: {descending: Block_Time}
      limit: {count: $limit}
      where: {
        Trade: {Currency: {MintAddress: {is: $token}}, Price: {gt: 0}
      }, 
        Transaction: {Result: {Success: true}}}
    ) {
      Block {
        allTime: Time
      }
      Trade {
        Account {
          Address
          Owner
        }
        Side {
          Type
          Account {
            Address
            Owner
          }
        }
        Price
        Amount
        Side {
          AmountInUSD
          Amount
        }
      }
    }
  }
}
"""

# ===================== NOT USED =====================
HISTORICAL_PRICE_AND_VOLUME_QUERY="""
query HistoricalPriceAndVolume($token: String!, $since: DateTime!, $interval_in: OLAP_DateTimeIntervalUnits!, $interval_count: Int!) {
  Solana {
    DEXTradeByTokens(
      where: {Trade: {Currency: {MintAddress: {is: $token}}}, Block: {Time: {since: $since}}, Transaction: {Result: {Success: true}}}
      orderBy: {ascending: Block_Time}
    ) {
      Block {
        Timefield: Time(interval: {in: $interval_in, count: $interval_count})
      }
      Trade {
        high: Price(maximum: Trade_Price)
        low: Price(minimum: Trade_Price)
        open: Price(minimum: Block_Time)
        close: Price(maximum: Block_Time)
        PriceInUSD
      }
      volume: sum(of: Trade_Amount)
      volume_in_usd: sum(of: Trade_Side_AmountInUSD)
      
    }
  }
}
"""

TOKEN_CREATION_QUERY="""
query MyQuery($token: String!) {
  Solana(network: solana) {
    Instructions(
      where: {Instruction: {Accounts: {includes: {Address: {is: $token}}}, Program: {Name: {is: "pump"}, Method: {is: "create"}}}}
    ) {
      Block{
        Time
      }
      Transaction {
        Signer
        Signature
      }
      Instruction {
        Accounts {
          Address
        }
      }
    }
  }
}
"""

LAST_N_TRANSACTIONS_QUERY="""
query LastNTransactionsByToken($token: String!, $n: Int!) {
  Solana {
    DEXTradeByTokens(
      where: {Trade: {Dex: {ProtocolName: {is: "pump"}}, Currency: {MintAddress: {is: $token}}}, Transaction: {Result: {Success: true}}}
      orderBy: {ascending: Block_Time}
      limit: {count: $n}
    ) {
      Block {
        Time
      }
      Trade {
        PriceInUSD
        Account {
          Address
          Owner
        }
        Amount
        Price
      }
      volume: sum(of: Trade_Amount)
      volume_in_usd: sum(of: Trade_Side_AmountInUSD)
      Transaction {
        Signature
      }
    }
  }
}
"""
# ===================== NOT USED =====================

File: supabase/functions/pumpfun_pipeline/utils/pumpfun.ts
import { SupabaseClient } from "jsr:@supabase/supabase-js@2";
import { getDateRange, parseDate } from "./time.ts";

const fetchBitQueryData = async (query: string) => {
  const headers = new Headers();
  headers.append("Content-Type", "application/json");
  headers.append("Authorization", `Bearer ${Deno.env.get("BITQUERY_TOKEN")}`);

  const requestOptions = {
    method: "POST",
    headers: headers,
    body: query,
  };

  const resp = await fetch("https://streaming.bitquery.io/eap", requestOptions);
  return await resp.json();
};

const fetchTopPumpFunTokens = async () => {
  const query = JSON.stringify({
    query: `query TopPumpFunTokensByPrice {
        Solana {
          DEXTrades(
            limitBy: {by: Trade_Buy_Currency_MintAddress, count: 1}
            limit: {count: 10 }
            orderBy: {descending: Trade_Buy_PriceInUSD}
            where: {Trade: {Dex: {ProtocolName: {is: "pump"}}, Buy: {Currency: {MintAddress: {notIn: ["11111111111111111111111111111111"]}}}}, Transaction: {Result: {Success: true}}}
          ) {
            Trade {
              Buy {
                PriceInUSD
                Currency {
                  Name
                  Symbol
                  MintAddress
                  Uri
                }
              }
            }
          }
        }
      }`,
    variables: {},
  });

  try {
    const result = await fetchBitQueryData(query);
    const tokens = await Promise.all(
      result.data.Solana.DEXTrades.map(async (trade) => {
        const { Name, Symbol, MintAddress, Uri } = trade.Trade.Buy.Currency;
        const PriceInUSD = trade.Trade.Buy.PriceInUSD;

        let metadata = {};
        try {
          const metadataResponse = await fetch(Uri);
          metadata = await metadataResponse.json();
        } catch (error) {
          console.error("Error fetching metadata:", error);
        }

        const richDataResponse = await fetch(
          `https://frontend-api.pump.fun/coins/${MintAddress}`,
        );

        const richData = await richDataResponse.json();

        // Extract or compute additional fields if needed
        const creator = richData.creator;
        const imageUri = richData.image_uri;
        const marketCap = richData.usd_market_cap;

        return {
          chain_id: 1,
          price: PriceInUSD,
          token_supply: 1000000000,
          mint_address: MintAddress,
          name: Name,
          symbol: Symbol,
          metadata: metadata,
          creator_address: creator,
          image_uri: imageUri,
          market_cap: marketCap,
          created_at: new Date(),
          updated_at: new Date(),
        };
      }),
    );

    return tokens;
  } catch (error) {
    console.error("Error fetching top pump fun tokens:", error);
    return [];
  }
};

const fetchPumpFunTokenTrades = async (data, startTime, endTime) => {
  const tokenTrades = [];
  for (const token of data) {
    const { id, mint_address } = token;
    const query = JSON.stringify({
      query:
        `query pumpfunTokenLatestTrades($mintAddress: String, $startTime: DateTime, $endTime: DateTime) {
        Solana {
          DEXTradeByTokens(
            orderBy: {descending: Block_Time}
            where: {
              Block: {
                Time: {
                  since: $startTime
                  till: $endTime
                }
              }
              Trade: {
                Currency: {
                  MintAddress: {is: $mintAddress}
                }
                Price: {gt: 0}
                Dex: {
                  ProtocolName: {is: "pump"}
                }
              }
              Transaction: {
                Result: {Success: true}
              }
            }
          ) {
            Block {
              allTime: Time
            }
            Trade {
              Side {
                Type
                Account {
                  Address
                }
              }
              PriceInUSD
              Amount
            }
          }
        }
      }`,
      variables: {
        mintAddress: mint_address,
        startTime: startTime,
        endTime: endTime,
      },
    });

    try {
      const result = await fetchBitQueryData(query);

      const trades = result.data.Solana.DEXTradeByTokens.map((trade) => {
        const { PriceInUSD } = trade.Trade;
        const { Type, Account } = trade.Trade.Side;
        const { Address } = Account;
        const ts = trade.Block.allTime;

        return {
          token_id: id,
          trade_price: PriceInUSD,
          buyer_wallet_address: Type === "buy" ? Address : null,
          seller_wallet_address: Type === "sell" ? Address : null,
          created_at: parseDate(ts),
        };
      });
      tokenTrades.push(...trades);
    } catch (error) {
      console.error(`Error fetching data for token ${token}:`, error);
    }
  }

  return tokenTrades;
};

const fetchPumpFunTokenCreatorHoldings = async (data) => {
  const tokenHolders = [];
  try {
    for (const token of data) {
      const { id, creator_address, mint_address } = token;

      const query = JSON.stringify({
        query:
          `query getTokenCreatorHolding($mintAddress: String, $walletAddress: String) {
          Solana {
            BalanceUpdates(
              orderBy: {descendingByField: "TotalHolding"}
              where: {BalanceUpdate: {Currency: {MintAddress: {is: $mintAddress}}, Account: {Owner: {is: $walletAddress}}}}
            ) {
              BalanceUpdate {
                Account {
                  Owner
                }
              }
              TotalHolding: sum(of: BalanceUpdate_Amount, selectWhere: {gt: "0"})
            }
          }
        }`,
        variables: {
          walletAddress: creator_address,
          $mintAddress: mint_address,
        },
      });

      const result = await fetchBitQueryData(query);

      if (result.data.Solana.BalanceUpdates.length == 0) {
        continue;
      }

      const holdings = result.data.Solana.BalanceUpdates[0].TotalHolding;
      tokenHolders.push({
        token_id: id,
        wallet_address: creator_address,
        holdings: holdings,
        created_at: new Date(),
        updated_at: new Date(),
      });
    }
  } catch (error) {
    console.error(`Error fetching data for token: `, error);
  }

  return tokenHolders;
};

const fillTokens = async (supabase: SupabaseClient) => {
  const tokens = await fetchTopPumpFunTokens();
  console.log("Fetched tokens: ", tokens);
  const { data, error } = await supabase
    .from("tokens")
    .upsert(tokens, {
      onConflict: ["chain_id", "mint_address"],
    })
    .select("id, chain_id, price, mint_address");

  if (error) {
    console.error("Error upserting tokens:", error);
    return false;
  }

  console.log("Upserted tokens: ", data);
  return true;
};

const fillTokenTrades = async (supabase: SupabaseClient) => {
  const { data: currentTokens, error: tokenFetchError } = await supabase
    .from("tokens")
    .select("id, mint_address");

  if (tokenFetchError) {
    console.error("Error fetching tokens:", tokenFetchError);
    return false;
  }

  const { startDate, endDate } = getDateRange();

  const tokenTrades = await fetchPumpFunTokenTrades(
    currentTokens,
    startDate,
    endDate,
  );
  const { data: tokenTradesUpsertData, error: tokenTradesUpsertError } =
    await supabase
      .from("token_trades")
      .upsert(tokenTrades, {
        onConflict: [
          "token_id",
          "buyer_wallet_address",
          "seller_wallet_address",
          "trade_price",
          "created_at",
        ],
      })
      .select("token_id, trade_price, created_at");
  if (tokenTradesUpsertError) {
    console.error("Error upserting token trades:", tokenTradesUpsertError);
    return false;
  }

  console.log(
    `Upserted token trades between ${startDate} and ${endDate}`,
  );
  return true;
};

const fillTokenCreators = async (supabase: SupabaseClient) => {
  const { data: currentTokens, error: tokenFetchError } = await supabase
    .from("tokens")
    .select("id, creator_address, mint_address");

  if (tokenFetchError) {
    console.error("Error fetching tokens:", tokenFetchError);
    return false;
  }

  const tokenCreators = await fetchPumpFunTokenCreatorHoldings(currentTokens);

  const { data: tokenCreatorsUpsertData, error: tokenCreatorsUpsertError } =
    await supabase
      .from("token_creators")
      .upsert(tokenCreators, {
        onConflict: ["token_id", "wallet_address"],
      })
      .select("token_id, holdings, updated_at");

  if (tokenCreatorsUpsertError) {
    console.error("Error upserting token holders: ", tokenCreatorsUpsertError);
    return false;
  }

  console.log("Upserted token creators!");
  return true;
};

export { fillTokenCreators, fillTokens, fillTokenTrades };
